File: examples/python/start-workflow/typecheck.py
# pyright: reportUnusedVariable=false
# ruff: noqa: F841

from __future__ import annotations

import collections.abc
import typing

import temporalio.api.workflowservice.v1.request_response_pb2
import temporalio.workflow

import start_workflow as output  # pyright: ignore[reportImplicitRelativeImport]
import start_workflow.models as output_models  # pyright: ignore[reportImplicitRelativeImport]

TASK_QUEUE = "demo-task-queue"


@temporalio.workflow.defn
class ExampleWorkflow:
    @temporalio.workflow.run
    async def run(self, customer_id: str) -> str:
        return customer_id


handle_task = output.start_workflow(
    workflow=ExampleWorkflow.run,
    input="customer-123",
    workflow_id="workflow-id",
    task_queue=TASK_QUEUE,
)
typed_handle_task: collections.abc.Awaitable[output.StartedWorkflow] = handle_task

named_handle_task: collections.abc.Awaitable[output.StartedWorkflow] = (
    output.start_workflow(
        workflow="ExampleWorkflow",
        workflow_id="workflow-id-named",
        task_queue=TASK_QUEUE,
    )
)

cancel_task: collections.abc.Awaitable[
    temporalio.workflow.NexusOperationHandle[
        temporalio.api.workflowservice.v1.request_response_pb2.RequestCancelWorkflowExecutionResponse
    ]
] = output.cancel_workflow(
    workflow_execution=output_models.WorkflowExecution(workflow_id="workflow-id"),
)


async def use_handle() -> None:
    handle = await handle_task
    await handle.cancel()
    restarted_handle_task: collections.abc.Awaitable[output.StartedWorkflow] = (
        handle.restart_workflow(
            workflow=ExampleWorkflow.run,
            task_queue=TASK_QUEUE,
        )
    )
    result_task: collections.abc.Awaitable[collections.abc.Sequence[typing.Any]] = (
        handle.get_result()
    )

File: examples/typescript/start-workflow/typecheck.ts
import type * as common from "@temporalio/common";
import {
  StartedWorkflow,
  StartWorkflowExecutionRequest,
  cancelWorkflow,
  startWorkflow,
} from "./output/index.ts";

async function exampleWorkflow(customerId: string): Promise<string> {
  return customerId;
}

const request: StartWorkflowExecutionRequest<typeof exampleWorkflow> = {
  workflow: exampleWorkflow,
  input: ["customer-123"],
  workflowId: "workflow-id",
  taskQueue: "demo-task-queue",
};

const handlePromise: Promise<StartedWorkflow> = startWorkflow(request);
const namedHandlePromise: Promise<StartedWorkflow> = startWorkflow({
  workflow: "ExampleWorkflow",
  workflowId: "workflow-id-named",
  taskQueue: "demo-task-queue",
});
void cancelWorkflow({
  workflowExecution: {
    workflowId: "workflow-id",
  },
});

async function useHandle(): Promise<void> {
  const handle = await handlePromise;
  await handle.cancel();
  const restartedHandlePromise: Promise<StartedWorkflow> = handle.restartWorkflow(
    exampleWorkflow,
    "demo-task-queue",
  );
  void restartedHandlePromise;
  const resultPromise: Promise<common.Payload[]> = handle.getResult();
  void resultPromise;
}

void namedHandlePromise;
void useHandle;

// @ts-expect-error missing workflow args for a callable workflow
startWorkflow({
  workflow: exampleWorkflow,
  workflowId: "missing-input",
  taskQueue: "demo-task-queue",
});

// @ts-expect-error workflow args must match the workflow callable
startWorkflow({
  workflow: exampleWorkflow,
  input: [7],
  workflowId: "bad-input",
  taskQueue: "demo-task-queue",
});
